src/sentry/ratelimits/leaky_bucket.py

```diff
@@ -0,0 +1,212 @@
+    last_drip: float = 0  # unix timestamp of the last drip
+    current_level: float = 0  # current level of the bucket
+    wait_time: float = 0  # seconds to wait until next request is allowed
+
+
+class LeakyBucketRateLimiter:
+    NAMESPACE = "leaky_bucket_limiter"
+
+    class LimitExceeded(Exception):
+        def __init__(self, info: LeakyBucketLimitInfo) -> None:
+            self.info = info
+
+    def __init__(self, burst_limit: int, drip_rate: int) -> None:
+        cluster_key = settings.SENTRY_RATE_LIMIT_REDIS_CLUSTER
+        self.client = redis.redis_clusters.get(cluster_key)
+        self.burst_limit = burst_limit
+        self.drip_rate = drip_rate
+
+    def redis_key(self, key: str) -> str:
+        return f"{self.NAMESPACE}:{key}"
+
+    def validate(self) -> None:
+        try:
+            self.client.ping()
+            self.client.connection_pool.disconnect()
+        except Exception as e:
+            raise InvalidConfiguration(str(e))
+
+    def use_and_get_info(self, key: str, timestamp: float | None = None) -> LeakyBucketLimitInfo:
+        if timestamp is None:
+            timestamp = time()
+
+        redis_key = self.redis_key(key)
+        try:
+            bucket_size, drip_rate, last_drip, current_level, wait_time = leaky_bucket_info(
+                [redis_key],
+                [self.burst_limit, self.drip_rate, timestamp],
+                client=self.client,
+            )
+            last_drip = float(last_drip)
+            current_level = float(current_level)
+            wait_time = float(wait_time)
+            info = LeakyBucketLimitInfo(bucket_size, drip_rate, last_drip, current_level, wait_time)
+            return info
+        except Exception:
+            logger.exception(
+                "Could not determine leaky bucket limiter state", dict(redis_key=redis_key)
+            )
+        # fail open
+        return LeakyBucketLimitInfo(self.burst_limit, self.drip_rate)
+
+    def is_limited(self, key: str, timestamp: float | None = None) -> bool:
+        return bool(self.use_and_get_info(key, timestamp).wait_time)
+
+    def get_bucket_state(self, key: str) -> LeakyBucketLimitInfo:
+        try:
+            last_drip, current_level = map(
+                lambda x: float(x or 0),
+                self.client.hmget(
+                    self.redis_key(key),
+                    ["last_drip", "current_level"],
+                ),
+            )
+        except Exception:
+            logger.exception("Could not get bucket state", dict(key=self.redis_key))
+            return LeakyBucketLimitInfo(self.burst_limit, self.drip_rate)
+
+        return LeakyBucketLimitInfo(
+            self.burst_limit,
+            self.drip_rate,
+            last_drip,
+            current_level,
+            max(0, (current_level - self.burst_limit) / self.drip_rate),
+        )
+
+    def context(self, key: str) -> ContextManager[LeakyBucketLimitInfo]:
```

Comment:
What is the benefit of having a context manager here, why not just use `is_limited` method for checking?

To me it seems like an added complexity, we need to handle possible exception that is being raised from it, there is no clean up that is being done after the context is exited, and it seems to me that it can all be achieved with just calling `limiter.is_limited(key)`, but maybe I am missing some very obvious reason